eva/optimizer/rules/rules.py

```diff
@@ -238,22 +241,44 @@ def __init__(self):
     def promise(self):
         return Promise.EMBED_FILTER_INTO_GET
 
-    def check(self, before: Operator, context: OptimizerContext):
-        # nothing else to check if logical match found return true
-        return True
+    def check(self, before: LogicalFilter, context: OptimizerContext):
+        # System supports predicate pushdown only while reading video data
+        predicate = before.predicate
+        lget: LogicalGet = before.children[0]
+        if predicate and lget.dataset_metadata.is_video:
+            # System only supports pushing basic range predicates on id
+            video_alias = lget.video.alias
+            col_alias = f"{video_alias}.id"
+            pushdown_pred, _ = extract_pushdown_predicate(predicate, col_alias)
+            if pushdown_pred:
+                return True
+        return False
 
     def apply(self, before: LogicalFilter, context: OptimizerContext):
         predicate = before.predicate
         lget = before.children[0]
-        new_get_opr = LogicalGet(
-            lget.video,
-            lget.dataset_metadata,
-            alias=lget.alias,
-            predicate=predicate,
-            target_list=lget.target_list,
-            children=lget.children,
+        # System only supports pushing basic range predicates on id
+        video_alias = lget.video.alias
+        col_alias = f"{video_alias}.id"
+        pushdown_pred, unsupported_pred = extract_pushdown_predicate(
```

Comment:
Shall we try avoiding doing the same computation in rule check and apply?

Reply:
I thought of it but the only way to do it is to keep the state across rule check and apply, which feels wierd to me and might creep in some bugs. Do you any other thoughts?

Reply:
I thought of it. But, the only way to avoid it is by keeping the state across check and apply which might creep in bugs.

Reply:
Btw, do you have any thoughts on it?

Reply:
We can either keep the state in rule or in optimizer context. I think both we need to clean up afterwards. To enforce the cleaning up, we may can implement a step in the base class, though not sure it is a good idea or not.

Reply:
I like the `optimizer context` idea. But it is too much engineering for now. I do not think the computation is expensive enough for us to spend time on this optimization.